krux/parser.py

```diff
@@ -0,0 +1,318 @@
+        )
+    else:
+        group.add_argument(
+            '--log-to-stdout',
+            default=False,
+            action='store_true',
+            env_var=False,
+            help='Log to stdout/stderr -- useful for debugging!',
+        )
+
+    return parser
+
+
+def add_stats_args(parser):
+    """
+    Add stats-related command-line arguments to the given parser.
+
+    :argument parser: parser instance to which the arguments will be added
+    """
+    group = get_group(parser=parser, group_name='stats', env_var_prefix=False)
+
+    group.add_argument(
+        '--stats',
+        default=False,
+        action='store_true',
+        help='Enable sending statistics to statsd.'
+    )
+    group.add_argument(
+        '--stats-host',
+        default=DEFAULT_STATSD_HOST,
+        help='Statsd host to send statistics to.'
+    )
+    group.add_argument(
+        '--stats-port',
+        default=DEFAULT_STATSD_PORT,
+        help='Statsd port to send statistics to.'
+    )
+    group.add_argument(
+        '--stats-environment',
+        default=DEFAULT_STATSD_ENV,
+        help='Statsd environment.'
+    )
+
+    return parser
+
+
+def add_lockfile_args(parser):
+    group = get_group(parser=parser, group_name='lockfile', env_var_prefix=False)
+
+    group.add_argument(
+        '--lock-dir',
+        default=DEFAULT_LOCK_DIR,
+        help='Dir where lock files are stored'
+    )
+
+    return parser
+
+
+def get_group(parser, group_name, env_var_prefix=None):
+    """
+    Return an argument group based on the group name.
+
+    This will return an existing group if it exists, or create a new one.
+
+    :argument parser: :py:class:`python3:argparse.ArgumentParser` to
+                      add/retrieve the group to/from.
+
+    :argument group_name: Name of the argument group to be created/returned.
+    """
+
+    # We don't want to add an additional group if there's already a 'logging'
+    # group. Sadly, ArgumentParser doesn't provide an API for this, so we have
+    # to be naughty and access a private variable.
+    groups = [g.title for g in parser._action_groups]
+
+    if group_name in groups:
+        group = parser._action_groups[groups.index(group_name)]
+    else:
+        group = parser.add_argument_group(title=group_name, env_var_prefix=env_var_prefix)
+
+    return group
+
+
+class KruxParser(ArgumentParser):
+    def add_argument_group(self, env_var_prefix, *args, **kwargs):
+        """
+        Creates a KruxGroup object that wraps the argparse._ArgumentGroup and creates a nice group of arguments
+        that should be considered together.
+
+        :param env_var_prefix: Prefix to use for the environment variables support of this group. If set to None,
+                               uses the title of the _ArgumentGroup that this is wrapping. If set to False,
+                               does not add any prefix.
+        :type env_var_prefix: str | bool
```

Comment:
Correct, it does not handle `True` properly. Because it also takes `str`. The use case that I was thinking of were following:
* `None`: Let it decide on its own
* `False`: Don't set any prefix
* `A_RANDOM_STRING`: This is what I want

Making it `True` instead of `None` would be more confusing, I think. 

Also, I am providing a `None` option for backward compatibility. We currently have existing arguments like `--stats` and `--stats-environment`. With this naming convention, it would be like `STATS_STATS` or `STATS_STATS_ENVIRONMENT`. I wanted to avoid those scenarios.

Reply:
I vote for breaking backward compatibility here. @slumos is correct that an empty prefix will lead to confusion and bugs.

The other alternative to breaking backward compatibility is special-casing those, which feels bad man.

Reply:
> Correct, it does not handle True properly. Because it also takes str.

Then bad values should probably be rejected early:
```python
if env_var_prefix is None:
    self._env_prefix = self.title + '_'
elif env_var_prefix is False:
    self._env_prefix = ''
elif isinstance(env_var_prefix, basestring):
    self._env_prefix = str(env_var_prefix) + '_'
else:
    raise TypeError('Parameter env_var_prefix must be None, True or a string, '
                    'but it is of type {}'.format(type(env_var_prefix).__name__))
```

Reply:
I’m still missing why we want the default mode to be to accept environment variables. This is what breaks backwards-compatibility.

If the default mode were to ignore env. vars.—as we do now—then it wouldn’t matter whether there’s a prefix defined or whether there are single-character arguments. It’s only when someone turns the env. vars. feature _on_ that this would matter, and I agree that they do matter.

This would require another argument: `accept_env_args`:
`KruxParser(…, accept_env_args=False, env_var_prefix=None)`

Reply:
To be clear, I’m not tying to bully us down this path. I haven’t heard a reason why we want this feature on by default, and I’m all👂🏼👂🏼. It’s possible I missed a convo or have wax in stdin.

Reply:
Two different discussion points in this thread:
1. Should we default to fall back to environment variables?
2. Should we let people omit `env_var_prefix`?

My arguments on each points:
1. This will cause some issues if your application has single character options like `-p`. The moment they upgrade `krux-stdlib`, their argparse will break. This is not a nice experience in minor version upgrade.  I agree with @dllopis-sfdc that we should default not to fall back to environment variable in this release and change that default in v3.0.
2. I am fine with forcing people to always use `env_var_prefix`. In that case, how should we handle generic arguments like `--stats-environment`? Are we OK with funny names like `STATS_STATS_ENVIRONMENT` or should these be special override cases? I personally prefer the latter.

Reply:
Special override cases for backward compatibility is my vote.